src/libstore/unix/build/derivation-builder.cc

```diff
@@ -671,23 +664,33 @@ static void replaceValidPath(const Path & storePath, const Path & tmpPath)
        tmpPath (the replacement), so we have to move it out of the
        way first.  We'd better not be interrupted here, because if
        we're repairing (say) Glibc, we end up with a broken system. */
-    Path oldPath = fmt("%1%.old-%2%-%3%", storePath, getpid(), rand());
-    if (pathExists(storePath))
+    Path oldPath;
+    
+    if (pathExists(storePath)) {
+        // why do we loop here?
+        // although makeTempPath should be unique, we can't 
+        // guarantee that.
+        do {
+            oldPath = makeTempPath(storePath, ".old");
+            // store paths are often directories so we can't just unlink() it
+            // let's make sure the path doesn't exist before we try to use it
+        } while (pathExists(oldPath));
         movePath(storePath, oldPath);
```

Comment:
file-system.hh also has `moveFile` which doesn't quite do this.
It has a guard for if the file exists but it does `copyFile` -- this seems to be only `mv`
(I think `moveFile` in file-system.hh is a bit mis-written since it does copy)